svm-transaction/src/instruction.rs

```diff
@@ -0,0 +1,24 @@
+use solana_sdk::instruction::CompiledInstruction;
+
+/// A non-owning version of [`CompiledInstruction`] that references
+/// slices of account indexes and data.
+// `program_id_index` is still owned, as it is a simple u8.
+#[derive(Debug, PartialEq, Eq, Clone)]
+pub struct Instruction<'a> {
```

Comment:
Any concern with naming this just `Instruction` and making it public? It's not a trait, but does `SVMInstruction` make sense here?

Reply:
done: 8950611ba95419a2b17ecc5151f36ce8c90e83ee

Reply:
didn't change the module names, but lmk if you think I should.

Reply:
Completely up to you. You could rename them for consistency with the others, but the modules prefixed with `svm_` are the traits, while these two are structs.

Reply:
Just lmk so I can reapprove when you're ready.

Reply:
We already have a `struct Instruction` in the repository with a similar usage:

https://github.com/anza-xyz/agave/blob/c338fbce1ac593a5d0001515257fc32055d2f3b8/sdk/program/src/instruction.rs#L333

Is this new struct going to supersede the existing `struct CompiledInstruction`?

Who's going to own the data for each instruction in your new strategy?

Reply:
The transaction type still owns the data, that's actually still required by this interface - rust won't let us return a reference to something we don't own (assuming `self` is only arg).

I could rename this struct to `CompiledInstruction`, since it is really a copy of that - I wasn't even aware there was just an `Instruction` type in sdk...it appears to be only used when **building** transactions, but not in the actual transaction types.

See this byte diagram here: https://apfitzge.github.io/posts/optimizing-scheduler-performance-accelerating-packet-to-transaction-conversion/#serialized-transactions

With the current `CompiledInstruction` it forces us to have owned Vecs for our account indexes and data. That's possible right now because our structure directly owns a Vec of these `CompiledInstruction`. But if we don't have that, and simply have offsets into an array of bytes, then we'd need some hack to create vecs from raw parts and then safely destroy them since we cannot actually free that memory.

Reply:
Inside of SVM, we only ever use slice-references to these. We never require its a vec nor modify them - so a non-owning version is more appropriate imo, and doesn't force us into a single design.

Reply:
In the short term, `CompileInstruction` will continue to exist, but your plan is to replace it with `TransactionView`, right?

> then we'd need some hack to create vecs from raw parts and then safely destroy them since we cannot actually free that memory.

I think this would lead to cleaner code. Trying zero-copy deserialization could be worth it. There exists a library for that (I haven't check if it useful for us, but perhaps stealing its ideas would be good) called [flat buffers](https://en.wikipedia.org/wiki/FlatBuffers).

Reply:
I think even long-term `CompiledInstruction` and `(Sanitized)VersionedTransaction` will continue to exist.
Those types are very useful in constructing txs, dev-facing tools for testing, etc, where deserialization speed does not matter as much.

The new transaction type, a simple serialized buffer, is not that convenient for building transactions, so I'm not sure we will ever fully deprecate the existing types.

In terms of flatbuffers, afaik it would have required us to use flatc for our serialization format as well.
The new tx type I'm working on will use the current `bincode` serialization format, but just be much faster than `bincode`s deserialization.

Reply:
Perhaps zero-copy deserialization for bincode is something worth looking into.

(The PR looks good - This is nothing for now).

Reply:
Yeah new deserialization is zero-copy. It is explicitly not supported by serde deserialization stuff, because alignment of types can get messed up in a serialized format, in the generic case.

In our specific case, the alignment of all types within our transaction is 1, so it cannot be off.